src/ports/postgres/modules/sample/balance_sample.py_in

```diff
@@ -0,0 +1,748 @@
+                    ) q2
+                ) q3
+            WHERE {row_number_col} <= {desired_count_col}
+        """.format(null_level_val=NULL_IDENTIFIER, **locals())
+    return subquery
+# ------------------------------------------------------------------------------
+
+
+def _get_with_replacement_subquery(schema_madlib, source_table,
+                                   source_table_columns, class_col,
+                                   actual_level_counts, desired_level_counts):
+    """ Return the query for sampling with replacement for specific class
+        levels (always used for oversampling, and used for undersampling if
+        with_replacement flag is set to TRUE).
+    """
+    if not desired_level_counts:
+        return ''
+
+    class_col_tmp = unique_string()
+    desired_count_col = unique_string()
+    actual_count_col = unique_string()
+    q1_row_no = unique_string()
+    q2_row_no = unique_string()
+
+    null_value_string = "'{0}'".format(NULL_IDENTIFIER)
+
+    desired_and_actual_level_counts = "VALUES " + \
+    ','.join("({0}, {1}, {2})".
+             format("'{0}'::text".format(k) if k else null_value_string,
+                v, actual_level_counts[k])
+             for k, v in desired_level_counts.items())
+    subquery = """
+            SELECT {source_table_columns}
+            FROM
+                (
+                    SELECT
+                         {class_col_tmp},
+                         generate_series(1, {desired_count_col}::int) AS _i,
+                         ((random()*({actual_count_col}-1)+1)::int) AS {q1_row_no}
+                    FROM
+                        ({desired_and_actual_level_counts})
+                            q({class_col_tmp}, {desired_count_col}, {actual_count_col})
+                ) q1,
+                (
+                    SELECT
+                        *,
+                        row_number() OVER(PARTITION BY {class_col}) AS {q2_row_no}
+                    FROM
+                         {source_table}
+                ) q2
+            WHERE {class_col_tmp} = coalesce({class_col}::text, '{null_level_val}') AND
+                  q1.{q1_row_no} = q2.{q2_row_no}
+        """.format(null_level_val=NULL_IDENTIFIER, **locals())
+    return subquery
+# ------------------------------------------------------------------------------
+
+def balance_sample(schema_madlib, source_table, output_table, class_col,
+                   class_sizes, output_table_size, grouping_cols,
+                   with_replacement, keep_null, **kwargs):
+    """
+    Balance sampling function
+    Args:
+        @param source_table       Input table name.
+        @param output_table       Output table name.
+        @param class_col          Name of the column containing the class to be
+                                  balanced.
+        @param class_sizes        Parameter to define the size of the different
```

Comment:
Discussions on changing param names should probably move to the JIRA, since the interface was kind of decided there. Other folks can comment on it too.